plutus-ir/src/Language/PlutusIR/Parser.hs

```diff
@@ -0,0 +1,344 @@
+{-# LANGUAGE FlexibleContexts #-}
+module Language.PlutusIR.Parser where
+
+import           Prelude                      hiding (fail)
+
+import           Control.Applicative          hiding (many, some)
+import           Control.Monad.State          hiding (fail)
+
+import qualified Language.PlutusCore          as PLC
+import qualified Language.PlutusCore.Constant as PLC
+import           Language.PlutusIR            as PIR
+import           Text.Megaparsec              hiding (State, parse)
+
+import           Data.ByteString.Internal     (c2w)
+import qualified Data.ByteString.Lazy         as BS
+import           Data.Char
+import           Data.Foldable
+import qualified Data.Map                     as M
+import qualified Data.Text                    as T
+import           Data.Word
+import           GHC.Natural
+
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer   as Lex
+
+data ParserState = ParserState { identifiers :: M.Map String PLC.Unique, counter :: PLC.Unique }
```

Comment:
Rather than have a counter here, I think you want to use the `Quote` monad or `QuoteT` which handles fresh name generation (essentially by having a counter inside it). You'll still need the map to lookup existing assignments, though.

Reply:
Yes I realise the PLC lexer doesn't do this, that's a bug arguably.

Reply:
Done, but I need an instance for `MonadQuote m => MonadQuote (ParsecT e s m)`, should I leave it orphaned or add it to `Quote.hs` and have to import `megaparsec` in `language-plutus-core`?

Reply:
Both options are fine to me. Now that we have a `megaparsec` parser of Plutus IR, I guess it's easier to maintain two similar parsers rather than two completely distinct parsers, so I'm in favor of replacing the current Plutus Core parser with something like what you have here (reusing parts would be great, of course) and then the `MonadQuote (ParsecT e s m)` instance belongs to `language-plutus-core`.